### DISCUSSION TITLE
Appimage local settings

### DISCUSSION THREAD

**Discussion by JonnyTech** [doublecmd/doublecmd#1218] [General]
I access DC from various Linux computers. Had been using the regular version but recently found that AppImage works better. Is it possible to use a local folder for its config path?

As per https://doublecmd.github.io/doc/en/commandline.html - `/mount-dir/mount-path/doublecmd-qt-latest-x86_64.AppImage --config-dir=/mount-dir/mount-path/config-dir/`

But the mount point varies from where I am accessing it from so do not want to hard-code it.

With the regular version, I can make a `settings` folder containing `doublecmd.inf` and the files there would be used automatically. Does similar exist with the AppImage?

The aim is to just execute `/mount-dir/mount-path/doublecmd-qt-latest-x86_64.AppImage` without the need of anything else and the local settings will be used.

```
Double Commander
Version: 1.2.0 alpha
Revision: 10882
Commit: 74c15bb
Build date: 2023/08/04
Lazarus: 2.2.6.0
FPC: 3.2.2
Platform: x86_64-Linux-qt5
OS version: MX 21.3 Wildflower

Widgetset library: Qt 5.15.2, libQt5Pas 5.6.1
```

**Comment by alexx2000** [comment]
> Had been using the regular version but recently found that AppImage works better.

What's better?

> Is it possible to use a local folder for its config path?

Create `doublecmd-qt-latest-x86_64` shell script near AppImage file, like:
```
#!/bin/sh

"$0".AppImage --config-dir="$0.settings" "$@"

```
and execute DC using it. DC will store settings in the `doublecmd-qt-latest-x86_64.AppImage.settings` directory near AppImage file in this case.

**Comment by JonnyTech** [comment]
> What's better?

Faster to start (probably because it is one file across the slow network rather than multiple small ones), no error if connection lost and try to access archive or anything else that requires a plugin (is AppImage loaded into memory?), some distros complained about dependancy (cannot remember which but I was surprised that AppImage did not), if I need to copy it locally then one single file is much faster.

My config file sets a few items and disables all histories so I shall not be writing to the remote share. I like to be as fully portable and trace-free as possible :)

**Comment by JonnyTech** [comment]
> "$0".AppImage --config-dir="$0.settings" "$@"

Clever trick using `$0`, it did not occur to me. Thanks, I shall try that tomorrow.

**Comment by JonnyTech** [comment]
AppImage files do some funny things...

I made the script and settings folder:

```
/media/sdb1/xfer# ls doublecmd-qt-latest-x86_64* -lad
-rwxr-xr-x 1 root root       56 Aug 11 12:42 doublecmd-qt-latest-x86_64*
-rwxr-xr-x 1 root root 15730704 Aug  5 14:40 doublecmd-qt-latest-x86_64.AppImage*
drwxr-xr-x 2 root root     8192 Aug 11 12:43 doublecmd-qt-latest-x86_64.settings/
```

The contents of the script:

```
/media/sdb1/xfer# cat doublecmd-qt-latest-x86_64
#!/bin/sh
"$0".AppImage --config-dir="$0.settings" "$@"
```

Then executed the script:
```
/media/sdb1/xfer# ./doublecmd-qt-latest-x86_64
SystemLocale en_US
SystemLanguage en
SystemEncoding UTF-8
DefaultSystemCodePage 65001
DefaultFileSystemCodePage 65001
DefaultRTLFileSystemCodePage 65001
ImageMagick: libMagickWand-7.Q16HDRI.so.10
Starting GuiMessageQueue
Starting Double Commander
Double Commander 1.2.0 alpha
Revision: 10882
Commit: 74c15bb
Build: 2023/08/04
Lazarus: 2.2.6.0
Free Pascal: 3.2.2
Platform: x86_64-Linux-qt5
System: Slackware Linux 
Desktop Environment: Xfce
Widgetset library: Qt 5.15.3, libQt5Pas 5.6.1
This program is free software released under terms of GNU GPL 2
(C)opyright 2006-2023 Alexander Koblov (alexx2000@mail.ru)
   and contributors (see about dialog)
Executable directory: /tmp/.mount_double2ZU4Gu/usr/lib64/doublecmd/
Configuration directory: /tmp/.mount_double2ZU4Gu/usr/doublecmd-qt-latest-x86_64.settings
Global configuration directory: /tmp/.mount_double2ZU4Gu/usr/lib64/doublecmd/settings/
Loading configuration from /tmp/.mount_double2ZU4Gu/usr/doublecmd-qt-latest-x86_64.settings/
Error loading configuration file /tmp/.mount_double2ZU4Gu/usr/doublecmd-qt-latest-x86_64.settings/favoritetabs.xml
SystemLocale en_US
SystemLanguage en
SystemEncoding UTF-8
DefaultSystemCodePage 65001
DefaultFileSystemCodePage 65001
DefaultRTLFileSystemCodePage 65001
PollThread: Start polling
PollThread: AddPoll 21
Creating PixmapManager
Error: Cannot save pixmaps cache [/tmp/.mount_double2ZU4Gu/usr/doublecmd-qt-latest-x86_64.settings/pixmaps.cache] : Unable to create file "/tmp/.mount_double2ZU4Gu/usr/doublecmd-qt-latest-x86_64.settings/pixmaps.cache"
Loading icon theme elementary Xfce
Theme elementary not found.
Theme Adwaita not found.
Theme gnome not found.
Loading icon theme Hicolor
Loading icon theme DCTheme
Loading icon theme DCTheme
Error loading configuration file /tmp/.mount_double2ZU4Gu/usr/doublecmd-qt-latest-x86_64.settings/shortcuts.scf
Error saving configuration file /tmp/.mount_double2ZU4Gu/usr/doublecmd-qt-latest-x86_64.settings/shortcuts.scf: Unable to create file "/tmp/.mount_double2ZU4Gu/usr/doublecmd-qt-latest-x86_64.settings/shortcuts~470353010.tmp"
Creating TFileSystemFileSource
TColumnsFileView.Create components
TColumnsFileView.Create components
FileSystemWatcher thread starting
PollThread: AddPoll 26
UDev: Begin monitoring
Detecting mounts through /proc/self/mounts
PollThread: AddPoll 27
frmMain.frmMainShow
```

My settings were not loaded.

So I tried the parameter directly, without the script:

```
/media/sdb1/xfer# ./doublecmd-qt-latest-x86_64.AppImage --config-dir=".doublecmd-qt-latest-x86_64.settings"
SystemLocale en_US
SystemLanguage en
SystemEncoding UTF-8
DefaultSystemCodePage 65001
DefaultFileSystemCodePage 65001
DefaultRTLFileSystemCodePage 65001
ImageMagick: libMagickWand-7.Q16HDRI.so.10
Starting GuiMessageQueue
Starting Double Commander
Double Commander 1.2.0 alpha
Revision: 10882
Commit: 74c15bb
Build: 2023/08/04
Lazarus: 2.2.6.0
Free Pascal: 3.2.2
Platform: x86_64-Linux-qt5
System: Slackware Linux 
Desktop Environment: Xfce
Widgetset library: Qt 5.15.3, libQt5Pas 5.6.1
This program is free software released under terms of GNU GPL 2
(C)opyright 2006-2023 Alexander Koblov (alexx2000@mail.ru)
   and contributors (see about dialog)
Executable directory: /tmp/.mount_doubley8TsNG/usr/lib64/doublecmd/
Configuration directory: /tmp/.mount_doubley8TsNG/usr/.doublecmd-qt-latest-x86_64.settings
Global configuration directory: /tmp/.mount_doubley8TsNG/usr/lib64/doublecmd/settings/
Loading configuration from /tmp/.mount_doubley8TsNG/usr/.doublecmd-qt-latest-x86_64.settings/
Error loading configuration file /tmp/.mount_doubley8TsNG/usr/.doublecmd-qt-latest-x86_64.settings/favoritetabs.xml
SystemLocale en_US
SystemLanguage en
SystemEncoding UTF-8
DefaultSystemCodePage 65001
DefaultFileSystemCodePage 65001
DefaultRTLFileSystemCodePage 65001
PollThread: Start polling
PollThread: AddPoll 21
Creating PixmapManager
Error: Cannot save pixmaps cache [/tmp/.mount_doubley8TsNG/usr/.doublecmd-qt-latest-x86_64.settings/pixmaps.cache] : Unable to create file "/tmp/.mount_doubley8TsNG/usr/.doublecmd-qt-latest-x86_64.settings/pixmaps.cache"
Loading icon theme elementary Xfce
Theme elementary not found.
Theme Adwaita not found.
Theme gnome not found.
Loading icon theme Hicolor
Loading icon theme DCTheme
Loading icon theme DCTheme
Error loading configuration file /tmp/.mount_doubley8TsNG/usr/.doublecmd-qt-latest-x86_64.settings/shortcuts.scf
Error saving configuration file /tmp/.mount_doubley8TsNG/usr/.doublecmd-qt-latest-x86_64.settings/shortcuts.scf: Unable to create file "/tmp/.mount_doubley8TsNG/usr/.doublecmd-qt-latest-x86_64.settings/shortcuts~1063003609.tmp"
Creating TFileSystemFileSource
TColumnsFileView.Create components
TColumnsFileView.Create components
FileSystemWatcher thread starting
PollThread: AddPoll 26
UDev: Begin monitoring
Detecting mounts through /proc/self/mounts
PollThread: AddPoll 27
frmMain.frmMainShow
```

Looks like the AppImage extracts itself to `/tmp/.mount_double??????/` and the log shows this folder: `Configuration directory: /tmp/.mount_double??????/usr/.doublecmd-qt-latest-x86_64.settings`

Is the `--config-di` argument not being passed on to DC correctly?

**Comment by alexx2000** [comment]
It's because you execute command using relative path `./doublecmd-qt-latest-x86_64`. Change the script to:
```
#!/bin/sh
EXE=$(realpath "$0")
"$EXE".AppImage --config-dir="$EXE.settings" "$@"
```
It will support both absolute and relative paths.

**Comment by JonnyTech** [comment]
Thanks, works perfectly.

**Comment by anpic** [comment]
also
```
#! /usr/bin/env sh
EXE=$(realpath "$0")
"$EXE".AppImage --config-dir="$EXE.settings" "$@"
```

**Comment by whbowers2** [comment]
I know this has been open for more than a few years, but I just  saw it today.  I found the perfect answer last week.  Create a directory under the one that your AppImage lives in with the exact same name as the AppImage and an extension of .config.  So for DoubleCommander.AppImage in an AppImages folder simply create DoubleCommander.AppImage.config in the AppImages folder.  That's it!  Everything else is managed automatically.  I put together a quick script file to run my AppImages.  It checks for the config directory and creates it if necessary, then runs the AppImage.  Its completely portable, doesn't care where the USB is mounted or anything else.

The entire script is a little rough in spots and not working properly, otherwise I'd share it.  However, the part that works with the config file is:

# Check for the AppImage's configuration directory XX.config, creating it if needed
# This will allow the AppImage to act just like a Windows Portable App in that it
# will keep localized data in the subdirectory of the AppName.config, which will
# persist between sessions, including copying or moving the AppImage and data to
# a different device - WHB - 2026 July 08  1820

strConfigDir="$1.config"
echo checking for "$strConfigDir in $PWD"
# Check if the directory exists
if [ -d "$strConfigDir" ]; then
    echo The configfuration folder "$strConfigDir" exists.
else
   mkdir "$strConfigDir"
    echo "creating folder $strConfigDir"
fi




Bill
